spot/reentrancy-demo.ipynb

```diff
@@ -0,0 +1,830 @@
+    "        x = self.norm_4_6(x)\n",
+    "        \n",
+    "        x = self.deconv_5_1(x)\n",
+    "        x = self.concat_5_3(self.c_crop_5_2(x_residual_3), x)\n",
+    "        x = self.conv_5_4(x)\n",
+    "        x = self.relu_5_5(x)\n",
+    "        x = self.norm_5_6(x)\n",
+    "        x = self.conv_5_7(x)\n",
+    "        x = self.relu_5_8(x)\n",
+    "        x = self.norm_5_9(x)\n",
+    "        \n",
+    "        x = self.deconv_6_1(x)\n",
+    "        x = self.concat_6_3(self.c_crop_6_2(x_residual_2), x)\n",
+    "        x = self.conv_6_4(x)\n",
+    "        x = self.relu_6_5(x)\n",
+    "        x = self.norm_6_6(x)\n",
+    "        x = self.conv_6_7(x)\n",
+    "        x = self.relu_6_8(x)\n",
+    "        x = self.norm_6_9(x)\n",
+    "        \n",
+    "        x = self.deconv_7_1(x)\n",
+    "        x = self.concat_7_3(self.c_crop_7_2(x_residual_1), x)\n",
+    "        x = self.conv_7_4(x)\n",
+    "        x = self.relu_7_5(x)\n",
+    "        x = self.norm_7_6(x)\n",
+    "        x = self.conv_7_7(x)\n",
+    "        x = self.relu_7_8(x)\n",
+    "        x = self.norm_7_9(x)\n",
+    "        \n",
+    "        x = self.conv_8_1(x)\n",
+    "        return x\n",
+    "\n",
+    "dataroot = Path('/mnt/segmented-bob-ross-images/')\n",
+    "dataset = BobRossSegmentedImagesDataset(dataroot)\n",
+    "dataloader = DataLoader(dataset, shuffle=True, batch_size=8)\n",
+    "\n",
+    "model = UNet()\n",
+    "model.cuda()\n",
+    "criterion = nn.CrossEntropyLoss()\n",
+    "optimizer = optim.Adam(model.parameters(), lr=0.5)\n",
+    "scheduler = optim.lr_scheduler.CosineAnnealingWarmRestarts(optimizer, T_0=32)\n",
+    "\n",
+    "for epoch in range(NUM_EPOCHS):\n",
+    "    losses = []\n",
+    "\n",
+    "    for i, (batch, segmap) in enumerate(dataloader):\n",
+    "        optimizer.zero_grad()\n",
+    "        \n",
+    "        batch = batch.cuda()\n",
+    "        segmap = segmap.cuda()\n",
+    "\n",
+    "        output = model(batch)\n",
+    "        loss = criterion(output, segmap)\n",
+    "        loss.backward()\n",
+    "        optimizer.step()\n",
+    "        scheduler.step()\n",
+    "\n",
+    "        curr_loss = loss.item()\n",
+    "        losses.append(curr_loss)\n",
+    "\n",
+    "    print(f'Finished epoch {epoch}.')\n",
+    "\n",
+    "torch.save(model.state_dict(), '50_net.pth')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**This training script is not reentrant.** Saving the model to disk is the very last thing this training script does. If the run executing this training script gets interrupted, all training progress the model made prior to termination is lost forever!\n",
+    "\n",
+    "It's unsafe to assume that the machine running the script will actually succeed in executing it from start to finish:\n",
+    "\n",
+    "* The machine might experience a hardware failure, forcing early termination.\n",
```

Comment:
remove extra 'be' 'it may randomly be forced into the REPAIRING state and self-terminate.'